shared/src/balancer_event_handler.rs

```diff
@@ -63,31 +67,62 @@ pub struct WeightedPoolBuilder {
 }
 
 impl WeightedPoolBuilder {
-    fn into_pool(self) -> Option<WeightedPool> {
-        if let (Some(pool_registration), Some(tokens_registration)) =
-            (self.pool_registration, self.tokens_registration)
-        {
-            return Some(WeightedPool {
+    async fn into_pool(
+        self,
+        weight_fetcher: &dyn NormalizedWeightFetching,
+    ) -> Result<WeightedPool> {
+        if let (Some(pool_registration), Some(tokens_registration)) = (
+            self.pool_registration.clone(),
+            self.tokens_registration.clone(),
+        ) {
+            return Ok(WeightedPool {
                 pool_id: pool_registration.pool_id,
                 pool_address: pool_registration.pool_address,
                 tokens: tokens_registration.tokens,
+                normalized_weights: weight_fetcher
```

Comment:
Has there been a discussion as to whether this should happen on event processing or lazily when the pool is actually used? A reason to do it here would be that we only have to do it once (it looks immutable in the contract). 
However, the purpose of the pool registry is mainly to "know" which pools exist and maybe we can save a bunch of queries (one roundtrip per pool on coldstart) if we instead fetch these infos batched when we also query the reserves.

For this it might also matter if/how we plan on handling non-weighted pools. Right now it seems like this registry is agnostic to the specializtion, but with this PR it would only work for WeightedPools (what happens if we pick up a more general pool creation even)? I'd prefer if this registry stays pool agnostic, but if we want to have it here it should be more up-front about which types of pools it can handle (and not issue a RPC call if it knows it cannot succeed).

Reply:
Very good point to bring up. I thought it would be nice to have this information right off the bat, but had not considered the fact that General pools are the only ones with weights. I guess I was assuming that the other pools would still contain this information although the weights are implied by the pool type and wouldn't need to be fetched. I'd be happy to keep the pools agnostic and move the weight fetching to the pool fetching when we get the reserves. This would only really incur one additional EVM query (depending on the pool type). 

I do recall @nlordell mentioning that we were only aiming to support `WeightedPools` initially (i.e. those with General Specialization). Maybe we should wait and see what he is thinking.

Reply:
I have taken your comment into consideration and created a new PR which addresses this. This new approach fetches the pool weights during poolFetching when seeking the pool's reserves. This then allows us to filter and separate for different pools until they are fully supported, meanwhile only constructing the public facing WeightedPool Structure at the time all of this information is known.

Please see here for potentially superseding changes https://github.com/gnosis/gp-v2-services/pull/702 (Note that these changes incorporate the work of both this PR and #698).

Reply:
I actually think we should keep this as is. The reason I say this is that it is that a general Balancer pool registry is not super useful without knowing exactly what kind of pool is being used. The `specialization` doesn't give enough information in that regard.

I checked what Balancer does in their subgraph <https://github.com/balancer-labs/balancer-subgraph-v2> and they listen for `PoolCreated` events on the `WeighedPoolFactory` instead of the `Vault` in order to get a list of weighted pools.

Reply:
This would be an argument for making the PoolRegistry practically a WeightedPool registry. However, I'd still question if we should do this roundtrip on every pool creation event or lazily (and then batched) in the dynamic state fetching logic?

Reply:
I would also prefer lazily, but I do think that the registry should be `WeightedPool` speific.

Reply:
I will rework this and use the weighted pool factory here.